src/main/java/innohack/gem/service/MatchService.java

```diff
@@ -0,0 +1,229 @@
+package innohack.gem.service;
+
+import com.google.common.collect.Lists;
+import innohack.gem.dao.IGEMFileDao;
+import innohack.gem.dao.IGroupDao;
+import innohack.gem.entity.GEMFile;
+import innohack.gem.entity.rule.Group;
+import innohack.gem.entity.rule.rules.Rule;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@Service
+public class MatchService {
+  private static HashMap<String, Collection<Group>> matchedFileGroupTable =
```

Comment:
Static (class-scope) is dangerous as it can easily leads to memory leaks and concurrency issues... any reason why we cannot reduce to runtime scope? (i.e. remove static)

Reply:
Not sure if spring container will treat service as singleton. Will spring singleton scope annotation help?

Reply:
> Static (class-scope) is dangerous as it can easily leads to memory leaks and concurrency issues... any reason why we cannot reduce to runtime scope? (i.e. remove static)

My bad, I think I should have put it in DAO. I was thinking of placing it in a static variable first before we persist it in data storage later on.

Reply:
sure, make sense